connect-inject/endpoints_controller.go

```diff
@@ -0,0 +1,429 @@
+					r.Log.Info("registering proxy service", "service", proxyServiceRegistration.Name)
+					err = client.Agent().ServiceRegister(proxyServiceRegistration)
+					if err != nil {
+						r.Log.Error(err, "failed to register proxy service with Consul", "consul-proxy-service-name", proxyServiceRegistration.Name)
+						return ctrl.Result{}, err
+					}
+				}
+			}
+		}
+	}
+
+	// Compare service instances in Consul with addresses in Endpoints. If an address is not in Endpoints, deregister
+	// from Consul. This uses endpointAddressMap which is populated with the addresses in the Endpoints object during
+	// the registration codepath.
+	if err = r.deregisterServiceOnAllAgents(serviceEndpoints.Name, serviceEndpoints.Namespace, true, endpointAddressMap); err != nil {
+		r.Log.Error(err, "failed to deregister service instances on all agents", "k8s-service-name", serviceEndpoints.Name, "k8s-namespace", serviceEndpoints.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// createServiceRegistrations creates the service and proxy service instance registrations with the information from the
+// Pod.
+func (r *EndpointsController) createServiceRegistrations(pod corev1.Pod, serviceEndpoints corev1.Endpoints) (*api.AgentServiceRegistration, *api.AgentServiceRegistration, error) {
+	// If a port is specified, then we determine the value of that port
+	// and register that port for the host service.
+	var servicePort int
+	if raw, ok := pod.Annotations[annotationPort]; ok && raw != "" {
+		if port, _ := portValue(&pod, raw); port > 0 {
+			servicePort = int(port)
+		}
+	}
+
+	// TODO: remove logic in handler to always set the service name annotation
+	// We only want that annotation to be present when explicitly overriding the consul svc name
+	// Otherwise, the Consul service name should equal the Kubernetes Service name.
+	// The service name in Consul defaults to the Endpoints object name, and is overridden by the pod
+	// annotation annotationService.
+	serviceName := serviceEndpoints.Name
+	if serviceNameFromAnnotation, ok := pod.Annotations[annotationService]; ok && serviceNameFromAnnotation != "" {
+		serviceName = serviceNameFromAnnotation
+	}
+
+	serviceID := fmt.Sprintf("%s-%s", pod.Name, serviceName)
+
+	meta := map[string]string{
+		MetaKeyPodName:         pod.Name,
+		MetaKeyKubeServiceName: serviceEndpoints.Name,
+		MetaKeyKubeNS:          serviceEndpoints.Namespace,
+	}
+	for k, v := range pod.Annotations {
+		if strings.HasPrefix(k, annotationMeta) && strings.TrimPrefix(k, annotationMeta) != "" {
+			meta[strings.TrimPrefix(k, annotationMeta)] = v
+		}
+	}
+
+	var tags []string
+	if raw, ok := pod.Annotations[annotationTags]; ok && raw != "" {
+		tags = strings.Split(raw, ",")
+	}
+	// Get the tags from the deprecated tags annotation and combine.
+	if raw, ok := pod.Annotations[annotationConnectTags]; ok && raw != "" {
+		tags = append(tags, strings.Split(raw, ",")...)
+	}
```

Comment:
@ishustava any thoughts on this? I'd be ok to remove it.

Reply:
It looks like the annotation was renamed in #141, and we've kept both names for backward compatibility. Since this is meant to be a refactor I think we can keep it. It seems relatively cheap to maintain.